osu.Game/Online/Multiplayer/MultiplayerClient.cs

```diff
@@ -175,33 +184,34 @@ public async Task JoinRoom(Room room, string? password = null)
             if (Room != null)
                 throw new InvalidOperationException("Cannot join a multiplayer room while already in one.");
 
+            Debug.Assert(room.RoomID != null);
+
             var cancellationSource = joinCancellationSource = new CancellationTokenSource();
+            await initRoom(room, r => JoinRoomInternal(room.RoomID.Value, password ?? room.Password), cancellationSource.Token).ConfigureAwait(false);
+        }
 
+        private async Task initRoom(Room room, Func<Room, Task<MultiplayerRoom>> initFunc, CancellationToken cancellationToken)
```

Comment:
Mild code quality concern, no action expected: I've gotten kinda wary of passing `Func<>`s in like this, and the nondescript variable name (`initFunc`) isn't helping me much. What I'd propose doing instead is:

- Extract the body of the innermost delegate here to a separate method (lines 200 through 228, call it `setUpJoinedRoomAsync()` or something)
- Half-inline this method in both call sites (would be something like

```csharp
await joinOrLeaveTaskChain.Add(async () =>
{
    await JoinRoomInternal(room.RoomID.Value, password ?? room.Password, cancellationSource.Token).ConfigureAwait(false);
    await setUpJoinedRoomAsync(); // add whatever parameters this needs to work
}, cancellationSource.Token).ConfigureAwait(false);
```

Reply:
I agree with this. Done in https://github.com/ppy/osu/pull/31637/commits/47ca5c90a5bada5733c89376916236b29c69467f